server: disable context shift

VJHack opened the pull request:
- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Low
  - [ ] Medium
  - [ ] High

This enables the `--no-context-shift` argument to be passed to the server.
The server will generate `n_predict` tokens such that `n_predict <=  n_ctx - n_tokens_prompt`.
If `n_tokens_prompts > n_ctx`, an error will be thrown and the slot is discarded.

Implements feature request #9390

VJHack committed:
allow disable context shift for sever

eskeletor97 commented on examples/server/server.cpp:
```diff
@@ -1940,6 +1947,12 @@ struct server_context {
+                        // context shift is disabled and prompt is too large - discard it
+                        if (!params.ctx_shift && slot.n_prompt_tokens > slot.n_ctx ){
+                            slot.release();
+                            send_error(slot, "Input is too large to process. Either disable context shift or increase context length. ", ERROR_TYPE_SERVER);
```
Shouldn't the error say enable context shift, since it's already disabled?

VJHack replied:
Yes, I made the change. Thanks for the correction.

ExtReMLapin commented:
n_ctx divided by slots available, right ?

VJHack committed:
Fixed error message to say  'enable context shift'

VJHack committed:
fixed server 200 null response when context is exceeded

VJHack committed:
changed error message wording

VJHack commented:
> n_ctx divided by slots available, right ?

slot.n_ctx for each slot is allocated by n_ctx divided by total number of slots;

ExtReMLapin commented:
Accoding to your commit message it returns 200, best would be [413](https://developer.mozilla.org/fr/docs/Web/HTTP/Status/413)

VJHack committed:
updated context shift error to ERROR_TYPE_INVALID_REQUEST

VJHack commented:
@ExtReMLapin Sorry if the commit message wasn't clear. In the commit that you're referring to, it was initially returning a 200 null response but I fixed it so it returns 500 with error message "Input is too large to process. Either enable context shift or increase the context length."

But now that you mention it, I think 400 is more appropriate because it matches what [OpenAI uses in their API response](https://community.openai.com/t/error-400-maximum-context-length-exceeded/931400). Just updated it to 400.

ExtReMLapin commented:
Thanks for the fix, I was about to open a PR to add 413 but if it's what openAI is doing, fair enough !

eskeletor97 commented on examples/server/server.cpp:
```diff
@@ -1940,6 +1946,12 @@ struct server_context {
                             continue;
                         }
+                        // context shift is disabled and prompt is too large - discard it
+                        if (!params.ctx_shift && (slot.n_prompt_tokens > slot.n_ctx) ){
```
I don't know if this leads to correct behavior. I found that if we do:
```
slot.n_prompt_tokens > slot.n_ctx
```
Then, it's possible to fall through the check down to prompt truncation which might be confusing for the user. Maybe we should change it to:
```
slot.n_prompt_tokens >= slot.n_ctx
```
Maybe someone more knowledgeable could chime in.

ngxson commented:
As I'm merging #9607 , I'll close this PR. Feel free to discuss on the other PR if you want to change something.
